Domain Search: Add Traintracks

umurkontaci opened the pull request:
This adds Traintracks for domain search in signup and domain search in Calypso.

Right now it treats signup and domains the same, they are sharing the code but we might differentiate them if it makes it easier.

@sirino Could you confirm this is in a shape that you can use?

sirino commented:
I used the calypso.live branch and https://mc.a8c.com/tracks/live to confirm that Traintracks events are produced and that all fields are present and have meaningful data. I got all the events I expected and the data looked good.

I noticed that you're using only ui_position and no fetch_position - I take it that that means that the two are the same? If so, it's ok to use only one.

The fetch_algo in the events is "domainsbot". Is there only one true "domainsbot" algorithm or should we append a version number or something to indicate which settings were used? Apart from that, everything looks great!

This is good to go from my side. Thanks a lot for adding Traintracks!

umurkontaci commented:
> I noticed that you're using only ui_position and no fetch_position - I take it that that means that the two are the same? If so, it's ok to use only one.

They are the same so far, we might look into changing that in the future.

klimeryk commented:
Oh, and forgot one more thing: maybe it'd be beneficial to log also the presence of the `Recommended`/`Best alternative` flags? They seem very effective and I'd imagine the suggestions that have them will be more popular/effective?

umurkontaci commented:
> Oh, and forgot one more thing: maybe it'd be beneficial to log also the presence of the Recommended/Best alternative flags? They seem very effective and I'd imagine the suggestions that have them will be more popular/effective?

/cc: @sirino how would you recommend logging the presence of these?

sirino commented:
> Oh, and forgot one more thing: maybe it'd be beneficial to log also the presence of the Recommended/Best alternative flags? They seem very effective and I'd imagine the suggestions that have them will be more popular/effective?
/cc: @sirino how would you recommend logging the presence of these?

Right, that would be an important information! I would suggest logging this as part of the recommendation results in a way that we can easily extract that information. For example, awesomedomain.blog#recommended or a similar format would work for us.

aidvu commented:
Might give you a skewed picture, as the nudges are psychological. You could randomly stick them, and they would have the same effect.

sirino commented:
I am sure those extra nudges will skew the results (even just the order will) but I think there's no better way of accounting for them than to start by tracking where they appear. I think with the recommendation result is the right place for that (and will make the analysis easier than, say, tracking it with the algorithm).

umurkontaci commented:
@delputnam @klimeryk I've rebased, added versioning and comment on the hash generating function.

Batch requests need to be handled accepted at Tracks first before we implement it.
I will leave deciding on what to do with `*.wordpress.com` and tagging things with Recommended labels to you.

klimeryk commented:
@delputnam, are you OK with taking over this PR? There's just a few small-ish things to fix (or at least investigate and promptly ignore for being to nitpicky). We should merge this PR soon, since it's quite old already.

delputnam commented:
@klimeryk Yep. I will update and try to get this one merged this week. I read about train tracks on Tuesday so, I'm probably an expert now. :-)

klimeryk commented on client/components/domains/domain-search-results/index.jsx:
```diff
@@ -148,7 +149,7 @@ var DomainSearchResults = React.createClass( {
 						railcarId={ `${ this.props.railcarSeed }-registration-suggestion-${ i }` }
 						uiPosition={ i }
-						fetchAlgo={ this.props.fetchAlgo }
+						fetchAlgo={ endsWith( suggestion.domain_name, 'wordpress.com' ) ? 'wpcom' : this.props.fetchAlgo }
```
While `wordpress` is a blacklisted word and should not be present in the domain name, super admins are exempt from this restriction, so it is technically possible to register, for example, `somerandomwordpress.com`, which will get flagged as `wpcom`, when it's not. Changing `'wordpress.com'` to `'.wordpress.com'` should fix that :)

delputnam replied:
Ah.  Good point.

klimeryk commented on client/components/domains/domain-registration-suggestion/index.jsx:
```diff
@@ -34,11 +34,18 @@ const DomainRegistrationSuggestion = React.createClass( {
 				ui_position: this.props.uiPosition,
 				fetch_algo: this.props.fetchAlgo,
-				rec_result: this.props.suggestion.domain_name,
+				rec_result: `${ this.props.suggestion.domain_name }#${ resultSuffix }`,
```
This causes `rec_result` to have `#` after the domain name always - even if there's no flag. So it becomes `example.com#` or if there's a flag: `example.com#recommended`.
@sirino, maybe we could just log the flag as a separate field? Suffixing it here seems like it's more limiting (we have to parse it for every result) than keeping it in a separate field (which could then be used to more easily filter the results?). Or am I understanding this incorrectly and this format actually does work for you better? :)

delputnam replied:
Oops! I moved the `#` into the `resultSuffix` variable.

This was the format that Sirin suggested using to add this information in the comment above. We don't have a way of adding arbitrary fields for a `_traintracks_render` event as far as I know.
